// common/database.js
var AWS = require('aws-sdk');

AWS.config.update({
    region: 'us-east-2'
});

var docClient = new AWS.DynamoDB.DocumentClient();

exports.getUser = getUser;
exports.setUser = setUser;
exports.deleteUser = deleteUser;

async function setUser(user) {
    let table = 'Users';
    var params = {
        TableName: table,
        Item: user
    };
    console.log('Adding a new item...');
    try {
        await docClient.put(params).promise();
        console.log('The item has been successfully added');
    } catch (err) {
        console.error(
            'Unable to add item. Error JSON:',
            JSON.stringify(err, null, 2)
        );
    }
}

async function getUser(id) {
    let table = 'Users';
    var params = {
        TableName: table,
        Key: {
            id
        }
    };
    try {
        const data = await docClient.get(params).promise();
        return data.Item;
    } catch (err) {
        console.error(err);
    }
}

function deleteUser(phone) {
    let table = 'Users';
    var params = {
        TableName: table,
        Key: {
            ID: phone
        }
    };
    docClient.delete(params, function(err, data) {
        if (err) {
            console.error(
                'Unable to delete item. Error JSON:',
                JSON.stringify(err, null, 2)
            );
        } else {
            console.log('DeleteItem succeeded:', JSON.stringify(data, null));
        }
    });
}
